Unit test user management

kohkaixun opened the pull request:
Under Private.test.js, the "it should render Spinner if axios request fails" (5th test) keeps throwing an error. Private.js doesn't have proper error validation to exit gracefully and that might be the issue. I feel like the test should be able to run and then show that test fails, however, whenever I run the test, the test doesn't seem to even run and I think the new error I created is thrown.

kohkaixun committed:
Install dependencies

kohkaixun committed:
Install Jest

kohkaixun committed:
Update gitignore

kohkaixun committed:
Remove gitignore cache

kohkaixun committed:
Update Jest test environment to node
Testing mongoose with Jest apparently doesn't work
with the default test environment of Jest. Online
recommendation to change to node.

kohkaixun committed:
Revert Jest test environment to default jsdom

kohkaixun committed:
Add Jest setup file

kohkaixun committed:
Add Jest setup file to package.json

kohkaixun committed:
Update Babel dependencies

kohkaixun committed:
Create Jest setup file

kohkaixun committed:
Install Jest dependencies

kohkaixun committed:
Create Private component test case for null auth

kohkaixun committed:
Add Private route test case for valid token

kohkaixun committed:
Add Private route test case for invalid token

kohkaixun committed:
Change all tokens to mockAuth

kohkaixun committed:
Add Private route test for valid token

kohkaixun committed:
Add Private route test for invalid token

kohkaixun committed:
Rename tests

kohkaixun committed:
Create Private route test for failed Axios request
Currently not working, need to fix test.
mockRejectedValueOnce is not working as expected.

kohkaixun committed:
Add Private component unit test for empty response
Test currently throws an error because this is not
caught in the original code. screen.debug() shows
that Spinner comoponent is rendered. However,
unsure if this is the correct way to test this.

kohkaixun committed:
Add Private route unit test for null res.data.ok
Similar to previous test for res = null, this test
throws an error. Unsure if this is the way to
test.

kohkaixun committed:
Update unit test for res.data.ok = null
Update: Test should pass because res.data.ok will
evaluate to false if it is null. Original code
will pass this test. Previous error was due to
misspelling of screen.debug() function.

kohkaixun committed:
Update unit tests for Private route
Total of 7 unit tests for Private.js.

First three tests check for null, valid and
invalid authentication tokens respectively. They
pass as expected.

Test 4 checks for when the axios.get call throws
an error. However, the test always throws an error
and causes all tests, including itself, to not
run. Unsure if this is the right way to test for
this scenario. Have skipped it for now.

Last three tests check for res, res.data and
res.data.ok being null. Test 5 and 6 throw errors
as they are not accounted for in original code and
causes all tests, including themselves, to not
run. Unsure if this is the right way to test for
these scenarios. Have skipped them for now.

kohkaixun committed:
Add comment to last test
Unsure if test 5, 6 and 7 should be added as they
seem to be ensuring that the code accepts a null
value as a valid response. However, the axios.get
function seems to either return the response or
throw an error. In this case, a null response
would indicate that the error is in the API
implementation and not the call. As such, should
these tests be added?

kohkaixun committed:
Added same tests for AdminRoute
As the code for AdminRoute.js and Private.js are
very similar, same tests have been added. However,
tests 4, 5 and 6 are still failing and still
unsure if to add tests 5, 6 and 7.

kohkaixun committed:
Rename unit tests for Private.js and AdminRoute.js

kohkaixun committed:
Merge branch 'main' into unit-test-user-management

kohkaixun committed:
Create unit test file for authController

kohkaixun committed:
Create updateProfile function unit test
Created unit tests for normal function and
different password lengths.

Normal function used password length of 8. Then
used boundary values 6 and 5 for tests 2 and 3
respectively.

kohkaixun committed:
Add unit tests for empty profile fields
3 different tests for different empty profile
fields.

- Empty email, password and phone fields
- Empty name and address fields
- All fields empty

kohkaixun committed:
Add unit tests for authController.js

kohkaixun committed:
Add unit test for when findByIdAndUpdate fails

kohkaixun committed:
Remove redundant import statement

kohkaixun committed:
Mock products for getOrdersController unit test

kohkaixun committed:
Fix unit tests for updateProfileController
All tests were failing dur to a redundant import
statement in the authController.test.js file that
imported beforeEach from node:test and resulted in
beforeEach not running at all. This caused
req.body to not be updated at all and all tests
were using the default req.body which was an empty
object.

kohkaixun committed:
Separate unit tests for different functions

kohkaixun committed:
Update updateProfileController unit tests

kohkaixun committed:
Update updateProfileController unit tests

kohkaixun committed:
Add mock users

kohkaixun committed:
Merge back unit test files for authController

eugenetangkj commented on client/src/components/Routes/AdminRoute.test.js:
```diff
@@ -0,0 +1,111 @@
+    useAuth: jest.fn()
+}));
+
+jest.mock('mongoose', () => ({
```
Why is `mongoose` being mocked? I dont see it being used in `AdminRoute.js`

eugenetangkj commented on client/src/components/Routes/AdminRoute.test.js:
```diff
@@ -0,0 +1,111 @@
+
+    it('should render Spinner if no auth token', async () => {
+        useAuth.mockReturnValue([null, jest.fn()]);
+        render(<AdminRoute />);
```
Do you want to expect that `axios.get` is called with a particular argument, for example `/api/v1/auth/admin-auth`?

eugenetangkj commented on client/src/components/Routes/Private.test.js:
```diff
@@ -0,0 +1,111 @@
+    useAuth: jest.fn()
+}));
+
+jest.mock('mongoose', () => ({
```
Again, why is `mongoose` mocked?

eugenetangkj commented on controllers/authController.updateProfileController.test.js:
```diff
@@ -0,0 +1,234 @@
+        expect(userModel.findByIdAndUpdate).toHaveBeenCalledTimes(1);
+        expect(userModel.findByIdAndUpdate).toHaveBeenCalledWith(req.user._id, updatedUser, { new: true });
+        expect(res.status).toHaveBeenCalledWith(200);
+        expect(res.send).toHaveBeenCalledTimes(1);
```
Not sure if you want to assert the content of the response, such as `success` and `message` attributes.

kohkaixun replied:
Will that make the test too brittle?

eugenetangkj commented on controllers/authController.updateProfileController.test.js:
```diff
@@ -0,0 +1,234 @@
+        await updateProfileController(req, res);
+        expect(userModel.findByIdAndUpdate).toHaveBeenCalledTimes(1);
+        expect(userModel.findByIdAndUpdate).toHaveBeenCalledWith(req.user._id, updatedUser, { new: true });
+        expect(res.status).toHaveBeenCalledWith(200);
```
Your test is testing 'Should keep old profile if new profile is empty'. However, your `expect` statements do not seem to be verifying whether the contents of the user are being changed or not.

kohkaixun replied:
I think updatedUser declared in line 182 is made from parsing the oldProfileUpdate declared in line 15. I just change the name of the updatedUser to oldProfileUpdate? Or is there some other logic error that I'm missing?

eugenetangkj commented on controllers/authController.updateProfileController.test.js:
```diff
@@ -0,0 +1,234 @@
+    });
+
+    // This test will fail as there is currently no validation for phone number in original code.
+    it("should not update profile with invalid phone number", async () => {
```
Would be good to leave a comment above the tests that you do not expect to pass so it's easier for us to read! 😄

eugenetangkj reviewed (commented):
Great work so far! Keep it up

kohkaixun committed:
Mock users, products and orders

kohkaixun committed:
Merge branch 'main' into unit-test-user-management

kohkaixun committed:
Merge

kohkaixun committed:
Comment out mocked mongoose as not used

kohkaixun committed:
Add expect axios.get calls

kohkaixun committed:
Add expect axios.get calls

kohkaixun committed:
Update updateProfileController unit tests
Updated tests according to Eugene's comments in
the PR. Awaiting the replies to some of the
questiions regarding some of the comments.

kohkaixun committed:
Include import statements

kohkaixun committed:
Add unit test for getOrdersController

kohkaixun committed:
Add unit test for no orders from user

kohkaixun committed:
Add unit test for when find throws error

kohkaixun committed:
Update find error unit test

kohkaixun committed:
Add unit test for first populate function error

kohkaixun committed:
Add unit test for second populate function error

kohkaixun committed:
Check success field in unit tests

kohkaixun committed:
Rename some unit tests

kohkaixun committed:
Update beforeEach for last two test suites

kohkaixun committed:
Update getOrdersController unit tests

kohkaixun committed:
Add unit test for getAllOrdersController

kohkaixun committed:
Add unit tests for getAllOrdersController
Added unit tests for empty list (no orders), error
thrown on find, first and second populate
functions and sort function.

kohkaixun committed:
Add unit test for orderStatusController

kohkaixun committed:
Add unit tests for orderStatusController
Added unit tests for req.params, req.body and
req.params.orderId are null

kohkaixun committed:
Add unit test case for orderId not found

kohkaixun committed:
Add uni tests for orderStatusController
Added unit tests for null order status and
findByIdAndUpdate error.

kohkaixun committed:
Remove unncessary comments

kohkaixun committed:
Remove redundant import

kohkaixun committed:
Add unit test for UserMenu

kohkaixun committed:
Add unit test cases for Dashboard.js

kohkaixun committed:
Add unit test for useAuth error

kohkaixun committed:
Simplify functions

kohkaixun committed:
Add initial render unit test for Profile page

kohkaixun committed:
Add user menu render into unit test

kohkaixun committed:
Add unit test for correct rendering of update

kohkaixun committed:
Merge unit test cases into one

kohkaixun committed:
Add unit tests for error field and useAuth error

kohkaixun committed:
Add unit test for getItem error

kohkaixun committed:
Add unit test for setItem error

kohkaixun committed:
Update unit tests for Profile

kohkaixun committed:
Install dependencies

kohkaixun committed:
Add unit tests for orderModel

kohkaixun committed:
Remove test not included in scope

kohkaixun committed:
Change mongo url

kohkaixun committed:
Add unit tests for UserModel

kohkaixun committed:
Add unit test for Orders

kohkaixun committed:
Update test name

kohkaixun committed:
Remove 3 similar tests

kohkaixun committed:
Merge null and empty string case together

kohkaixun committed:
Separate unit test into file

kohkaixun committed:
Separate unit test into file

kohkaixun committed:
Separate unit test into file

kohkaixun committed:
Remove authController.test.js

kohkaixun committed:
Create separate unit test file for getAllOrdersController

kohkaixun committed:
Create separate unit test file for getOrdersController

kohkaixun committed:
Create separate unit test file for orderStatusController

kohkaixun committed:
Create separate unit test file for updateProfileController

kohkaixun committed:
Update unit test for Private

kohkaixun committed:
Update unit test for Dashboard

kohkaixun committed:
Reduce unit tests for Profile

kohkaixun committed:
Update unit test for orderModel

kohkaixun committed:
Reduce unit test for userModel

kohkaixun committed:
Merge branch 'main' into unit-test-user-management

eugenetangkj commented on client/src/pages/user/Dashboard.test.js:
```diff
@@ -0,0 +1,39 @@
+    it("should not render user information if auth is null", () => {
+        useAuth.mockReturnValue([null]);
+        render(<Dashboard />);
+        expect(screen.queryByText(name)).not.toBeInTheDocument();
```
Hmmm... okay, but maybe another approach would have been to check that the h3 elements are not rendered with any content.

eugenetangkj commented on client/src/pages/user/Profile.test.js:
```diff
@@ -0,0 +1,153 @@
+        fireEvent.change(phoneInput, { target: { value: updatedUserProfile.user.phone } });
+        fireEvent.change(addressInput, { target: { value: updatedUserProfile.user.address } });
+
+        expect(nameInput.value).toBe(updatedUserProfile.user.name);
```
Hmmm, this is checking whether the fields is editable + whether the updating of profile is correct. -> 2 different things.

But okay for this round ba

eugenetangkj reviewed (approved):
Thanks! Looks good, left some comments for your reference.
